src/Sarif.Viewer.VisualStudio/ErrorListService.cs

```diff
@@ -0,0 +1,155 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+
+using Microsoft.CodeAnalysis.Sarif.Driver;
+using Microsoft.CodeAnalysis.Sarif.Driver.Sdk;
+using Microsoft.CodeAnalysis.Sarif.Sdk;
+
+namespace SarifViewer
+{
+    public class ErrorListService
+    {
+        public static ErrorListService Instance = new ErrorListService();
+
+        public static void ProcessSarifLog(ResultLog resultLog)
+        {
+            foreach (RunLog runLog in resultLog.RunLogs)
+            {
+                Instance.WriteRunLogToErrorList(runLog);
+            }
+
+            SarifTableDataSource.Instance.BringToFront();
+        }
+
+        private ErrorListService()
+        {
+            this.documentToLineIndexMap = new Dictionary<string, NewLineIndex>();
+        }
+
+        private Dictionary<string, NewLineIndex> documentToLineIndexMap;
+
+        private IRuleDescriptor GetRule(RunLog runLog, string ruleId)
+        {
+            if (runLog.RuleInfo == null)
+            {
+                return null;
+            }
+
+            foreach (IRuleDescriptor ruleDescriptor in runLog.RuleInfo)
+            {
+                if (ruleDescriptor.Id == ruleId) { return ruleDescriptor; }
+            }
+
+            throw new InvalidOperationException();
```

Comment:
I took a look at the spec and section 9.5.2 already makes this clear. If you've got a formattedMessage that refers by ruleId to a formatSpecifier, it must be present in the ruleInfo data.

9.5.2. specifierId property

A formattedMessage object must contain a property named specifierId, whose value is a string that identifies the format string used to format the message that describes this result. The value of specifierId must correspond to one of the names in the set of name/value pairs contained in the formatSpecifiers property (§9.3.8) of the ruleInfo object (§9.3) whose id property (§9.3.3) matches the ruleId property (§6.9.2) of this issue.